### ISSUE TITLE
connected_components: respect diagonal values, exclude values which are never referenced

### ISSUE THREAD

**Issue by nschloe** [scipy/scipy#14391]
Right now, `connected_components` does not take into account the values on the diagonal of the input matrix. Whether you supply an all-zero matrix or the identity matrix of the same size, you always get each node index in its own component:
```python
from scipy.sparse.csgraph import connected_components

A = np.array([[0]])

num_cc, cc = connected_components(A)
print(num_cc, cc)
```
```python
1 [0]
```
I now have a use case where there are indices which are never referenced in the entire graph, and I would like them not to be counted anwhere. I would like the above snippet to return
```python
1 [None]   # or [np.nan], or -1
```
If I see correctly, this is currently not possible.

Considering that the values on the diagonal doesn't actually do anything, it seems to be made for this use case. 
I would suggest to add a flag (default `False` of course, to not break backwards-compatibility) which yields the following results:
```python
[[0]] -> 0 [None]
[[1]] -> 1 [0]
[[0, 0], [0, 0]] -> 0, [None, None]
[[1, 0], [0, 0]] -> 1, [0, None]
[[0, 0], [0, 1]] -> 1, [None, 0]
[[1, 0], [0, 1]] -> 2, [0, 1]
[[0, 1], [0, 0]] -> 1, [0, 0]
[[0, 1], [1, 0]] -> 1, [0, 0]
```

**Comment by rkern** [comment]
I don't think this meshes with the usual semantics of adjacency matrices or [standard definitions of connected components](https://en.wikipedia.org/wiki/Component_(graph_theory)). A graph is a pair of sets, one for nodes and one for edges. An adjacency matrix encodes a graph by assigning each node to a row/column and nonzero entries for the edges. The mere presence of the row/column means that the node is in the graph, even if there are no edges that start/end on that node. Nonzero entries on the diagonal mean that there are edges that loop back onto the node. Zero entries on the diagonal only mean that there are no self-looping edges for that node. Nodes with no edges each have their own component.

For the connected components computation, each node represented in the graph (i.e. a row/column exists) must appear in one of the connected components even if there are no edges corresponding to it. If you want to ignore all 1-node components that don't have a self-looping edge, I think that's something that should be done as post-processing.

**Comment by nschloe** [comment]
> For the connected components computation, each node represented in the graph (i.e. a row/column exists) must appear in one of the connected components even if there are no edges corresponding to it.

Well, I have a use case where this is _not_ desired. The post-processing necessary is
```python
_, cc = connected_components(M)
num_cc = np.sum(np.bincount(cc) > 1)
```
It for large matrices `M`, the post-processing takes about 8% of the `connected_components` computation, so it can probably be neglected. MWE:
```python
import numpy as np
from scipy.sparse import coo_matrix
from scipy.sparse.csgraph import connected_components
import meshzoo
import meshplex
import time


points, cells = meshzoo.ngon(6, 500)

mesh = meshplex.MeshTri(points, cells)

boundary_edges = mesh.edges["points"][mesh.is_boundary_edge]
n = np.max(boundary_edges) + 1
val = np.ones(len(boundary_edges), dtype=int)
t0 = time.time()
M = coo_matrix((val, boundary_edges.T), shape=(n, n))

t1 = time.time()
_, cc = connected_components(M)
t2 = time.time()
num_cc = np.sum(np.bincount(cc) > 1)
t3 = time.time()

print(f"time coo_matrix:           {t1 - t0:.3e}")
print(f"time connected_components: {t2 - t1:.3e}")
print(f"time sum-bincount:         {t3 - t2:.3e}")
```
```
time coo_matrix:           2.046e-04
time connected_components: 7.188e-04
time sum-bincount:         5.841e-05
```
